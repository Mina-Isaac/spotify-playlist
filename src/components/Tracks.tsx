import React, { useEffect } from "react";
import useInfiniteScroll from "../utilities/useInfiniteScroll";
import { observer } from "mobx-react";
import { useMst } from "../tree/Root";
import { useParams, useLocation } from "react-router-dom";
import { prepareData } from "../utilities/tools";

const Tracks = () => {
  const root = useMst();
  const list = root.Tracks.items.map((item) => item.track.artists).flat();
  const setIsFetching = useInfiniteScroll(moreData)[1];
  const { id } = useParams();
  const { state: playlistName } = useLocation();
  const loading = root.Tracks.loading;

  useEffect(() => {
    root.Tracks.fetchNext(id);
    return () => root.Tracks.clearTracks();
  }, []);

  function moreData() {
    root.Tracks.fetchNext();
    setIsFetching(false);
  }

  if (list.length === 0) {
    return <h1>Loading...</h1>;
  }

  return (
    <>
      <h1>Artists in {playlistName}</h1>
      <ul className="list">
        {prepareData(list).map((item, i) => (
          <li className="card" key={i}>
            <div className="txt">
              <span className="primary-txt">{item.name}</span>
              <span className="secondery-txt">{item.type}</span>
            </div>
          </li>
        ))}
      </ul>
      {loading && <span className="primary-text">Loading...</span>}
    </>
  );
};

export default observer(Tracks);
