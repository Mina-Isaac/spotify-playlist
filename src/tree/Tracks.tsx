import { types, flow, Instance } from "mobx-state-tree";
import appService from "./service";

interface ITrack extends Instance<typeof track> {}

const artist = types.model({
  id: types.identifier,
  name: types.string,
  type: types.string,
});

const track = types.model({
  track: types.model({
    artists: types.array(artist),
  }),
});

const Tracks = types
  .model({
    items: types.array(track),
    next: types.union(types.string, types.null),
    total: types.number,
    loading: types.boolean,
  })
  .actions((self) => {
    const fetchNext = flow(function* (id?) {
      if (self.next === null) return;
      self.loading = true;
      try {
        const newlyFetched = yield appService.fetchTracks(id, self.next);
        self.next = newlyFetched.next;
        (newlyFetched.items as Array<ITrack>).forEach((item) =>
          self.items.push(item)
        );
        self.loading = false;
      } catch (error) {
        console.error("Failed to fetch projects", error);
        self.loading = false;
      }
      return;
    });
    const clearTracks = () => {
      self.items.clear();
      self.loading = true;
      self.next = "";
      self.total = 0;
    };
    return { fetchNext, clearTracks };
  });

export default Tracks;
