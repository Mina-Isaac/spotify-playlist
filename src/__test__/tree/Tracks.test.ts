import Tracks from "../../tree/Tracks";
import { getSnapshot, onPatch } from "mobx-state-tree";

const sample = {
  loading: false,
  next: "some url",
  items: [
    {
      track: {
        artists: [{ id: "some id", name: "some name", type: "some type" }],
      },
    },
  ],
  total: 24,
};

it("should create a model instance", () => {
  const item = Tracks.create(sample);
  expect(getSnapshot(item)).toMatchSnapshot();
});

it("reset model", () => {
  const item = Tracks.create(sample);
  const patches: any = [];

  onPatch(item, (patch) => {
    patches.push(patch);
  });

  item.clearTracks();
  expect(patches).toMatchSnapshot();
});
