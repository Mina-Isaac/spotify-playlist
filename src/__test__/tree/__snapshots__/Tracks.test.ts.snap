// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reset model 1`] = `
Array [
  Object {
    "op": "remove",
    "path": "/items/0",
  },
  Object {
    "op": "replace",
    "path": "/loading",
    "value": true,
  },
  Object {
    "op": "replace",
    "path": "/next",
    "value": "",
  },
  Object {
    "op": "replace",
    "path": "/total",
    "value": 0,
  },
]
`;

exports[`should create a model instance 1`] = `
Object {
  "items": Array [
    Object {
      "track": Object {
        "artists": Array [
          Object {
            "id": "some id",
            "name": "some name",
            "type": "some type",
          },
        ],
      },
    },
  ],
  "loading": false,
  "next": "some url",
  "total": 24,
}
`;
